.github/workflows/release-cycle.yml

```diff
@@ -0,0 +1,213 @@
+name: Release Cycle
+
+on:
+  push:
+    branches:
+      - release-v*
+  workflow_dispatch: {}
+
+concurrency: ${{ github.workflow }}-${{ github.ref }}
+
+jobs:
+  check_state:
+    environment: Release Cycle
+    name: Check state
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up environment
+        uses: ./.github/actions/setup
+      - id: changeset_state
+        name: Read Changeset state
+        run: |
+          echo "changesets=$(node scripts/release/read-changeset-state.js changesets length)" >> $GITHUB_OUTPUT
+          echo "prestate=$(node scripts/release/read-changeset-state.js preState mode)" >> $GITHUB_OUTPUT
+          echo "packageJsonVersion=$(node -p -e "require('./package.json').version")" >> $GITHUB_OUTPUT
+      - id: check_pr_back
+        if: contains(github.ref_name, 'release-v') && github.event_name == 'push' && steps.changeset_state.outputs.prestate == 'undefined' && !contains(steps.changeset_state.outputs.packageJsonVersion, '-rc.') && steps.changeset_state.outputs.changesets == '0'
+        name: Check if PR back to master is required
+        uses: actions/github-script@v6
+        with:
+          result-encoding: string
+          script: |
+            const { data: PRs } = await github.rest.pulls.list({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              head: 'merge/${{ github.ref_name }}',
+              base: 'master',
+              state: 'open'
+            });
+
+            return (PRs.length === 0).toString();
+
+    outputs:
+      start_release_candidate: ${{ github.ref_name == 'master' && github.event_name == 'workflow_dispatch' }}
+      changesets_pr: ${{ contains(github.ref_name, 'release-v') && github.event_name == 'push' }}
+      promote_to_final: ${{ contains(github.ref_name, 'release-v') && github.event_name == 'workflow_dispatch' }}
+      publish: ${{ contains(github.ref_name, 'release-v') && github.event_name == 'push' && steps.changeset_state.outputs.changesets == '0' }}
+      pr_back: ${{ steps.check_pr_back.outputs.result == 'true' }}
+
+  start_release_candidate:
+    needs: check_state
+    environment: Release Cycle
+    name: Start new release branch with release candidate
+    if: needs.check_state.outputs.start_release_candidate == 'true'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          token: ${{ secrets.GH_TOKEN }} # Needed so release cycle triggers when pushing
```

Comment:
For the record:

- We removed the `GH_TOKEN` but started relying on a dispatch in [this commit](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3915/commits/3e8e900eec728ec6d7e295c2b5c1c5d2f09db8cf).
- We decided not to use `workflow_run: completed` since it'll trigger an infinite loop
- We went for setting `start` and `promote` workflows as dependencies of `changesets` so we remove the dispatches in [this commit](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3915/commits/4484b77beed493b7193709891d50331e8641fb0d)